File: dynamic_programming/subarray_and_subsequence/53_maximum_subarray.cc
#include <algorithm>
#include <vector>

// 2024/01/04: 完全自己写出dp方法
int MaxSubArray(std::vector<int>& nums) {
  int n = nums.size();
  // dp[i]: 以i结尾的连续数组的最大值.
  std::vector<int> dp(n, 0);
  dp[0] = nums[0];
  int res = nums[0];
  for (int i = 1; i < n; ++i) {
    dp[i] = nums[i] + std::max(dp[i - 1], 0);
    res = std::max(res, dp[i]);
  }
  return res;
}

File: dynamic_programming/subarray_and_subsequence/53_maximum_subarray.py
from typing import List


# 2026/05/08: 直接看答案.
# 要点是要能想到要 DP 做.
def max_sub_array(nums: List[int]) -> int:
    dp = [0] * len(nums)
    dp[0] = nums[0]
    for i in range(1, len(nums)):
        dp[i] = max(dp[i - 1], 0) + nums[i]
    return max(dp)


nums = [1]
print(max_sub_array(nums))
